consumer/file.go

```diff
@@ -280,8 +296,172 @@ func (cons *File) Consume(workers *sync.WaitGroup) {
+}
+
+func (source *sourceFile) getRealFileName() string {
+	baseFileName, err := filepath.EvalSymlinks(source.fileName)
+	if err != nil {
+		baseFileName = source.fileName
+	}
+
+	baseFileName, err = filepath.Abs(baseFileName)
+	if err != nil {
+		baseFileName = source.fileName
+	}
+
+	return baseFileName
+}
+
+func newSourceFile(conf core.PluginConfigReader) (sourceFile, error) {
+	source := sourceFile{}
+	err := conf.Configure(&source)
+	if err != nil {
+		return source, err
+	}
+
+	return source, nil
+}
+
+// -- seeker --
+
+type seeker struct {
+	seek     int
+	onRotate int
+	offset   int64
+}
+
+func newSeeker(conf core.PluginConfigReader) seeker {
+	switch strings.ToLower(conf.GetString("DefaultOffset", fileOffsetEnd)) {
+	default:
+		fallthrough
+	case fileOffsetEnd:
+		return seeker{
+			seek:     io.SeekEnd,
+			onRotate: io.SeekCurrent,
+			offset:   io.SeekStart,
+		}
+
+	case fileOffsetStart:
+		return seeker{
+			seek:     io.SeekCurrent,
+			onRotate: io.SeekCurrent,
+			offset:   io.SeekStart,
+		}
+	}
+}
+
+// -- watcher --
+
+type watcher struct {
+	logger logrus.FieldLogger
+	source *sourceFile
+	read   func()
+
+	done chan int
+}
+
+func newWatcher(logger logrus.FieldLogger, source *sourceFile, readFunction func()) *watcher {
+	watcher := watcher{}
+
+	watcher.logger = logger
+	watcher.source = source
+	watcher.read = readFunction
+
+	return &watcher
+}
+
+func (w *watcher) close(watcher *fsnotify.Watcher) {
+	err := watcher.Close()
+	if err != nil {
+		w.logger.Fatal(err)
+	}
+}
+
+func (w *watcher) Watch(buffer *tio.BufferedReader, sendFunction func(data []byte)) {
+	// init
+	w.done = make(chan int)
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		w.logger.Fatal(err)
+	}
+	defer w.close(watcher)
+
+	// watcher loop in subroutine
+	go w.watchLoop(watcher)
+
+	// busy loop for shutdown
+	for w.source.state != fileStateDone {
+		time.Sleep(time.Second)
+	}
+
+	w.logger.Debug("shutdown file watcher ..")
+	close(w.done)
+}
+
+func (w *watcher) watchLoop(watcher *fsnotify.Watcher) {
+outerLoop:
+	for {
+		if _, err := os.Stat(w.source.realFileName); os.IsNotExist(err) {
+			w.logger.WithField("file", w.source.realFileName).
+				Warning("watched file not exists. retry in 3sec ..")
+			time.Sleep(3 * time.Second)
+			continue outerLoop // retry
+		}
+
+		// read current file state before watching
+		w.read()
+
+		err := watcher.Add(w.source.realFileName)
+		if err != nil {
+			w.logger.Error("error during adding watcher: ", err)
+			time.Sleep(3 * time.Second)
+			continue outerLoop // retry
+		}
+
+	innerLoop:
+		for {
+			select {
+			case event := <-watcher.Events:
+				if event.Op&fsnotify.Write == fsnotify.Write {
+					w.logger.Debug("modified file: ", event.Name)
+					w.read()
+					break
+				}
+
+				if event.Op&fsnotify.Rename == fsnotify.Rename || event.Op&fsnotify.Remove == fsnotify.Remove {
+					w.logger.WithField("event", event).Debug("file renamed/removed: ", event.Name)
+					break innerLoop
```

Comment:
Better name? Maybe event loop ..?
IMO it's better to say what "action" you want to break